src/InsertCapablePdoTrait.php

```diff
@@ -0,0 +1,236 @@
+     *
+     * @return PDOStatement The executed PDO statement.
+     */
+    protected function _insert($records)
+    {
+        $processedRecords = $this->_preProcessRecords($records, $valueHashMap);
+
+        $query = $this->_buildInsertSql(
+            $this->_getSqlInsertTable(),
+            $this->_getSqlInsertColumnNames(),
+            $processedRecords,
+            $valueHashMap
+        );
+
+        $statement = $this->_executePdoQuery($query, array_flip($valueHashMap));
+
+        return $statement;
+    }
+
+    /**
+     * Pre-processes the list of records.
+     *
+     * @since [*next-version*]
+     *
+     * @param array[]|ArrayAccess[]|stdClass[]|ContainerInterface[]|Traversable $records      A list of records.
+     * @param array                                                             $valueHashMap A hash-to-value map
+     *                                                                                        reference to which new
+     *                                                                                        hash-value pairs are
+     *                                                                                        written.
+     *
+     * @throws ContainerExceptionInterface If an error occurred while reading from a record's container.
+     *
+     * @return array The pre-processed record data list, as an array of record data associative sub-arrays.
+     */
+    protected function _preProcessRecords($records, &$valueHashMap = [])
+    {
+        // Initialize variable, in case it was declared implicitly during the method call
+        if ($valueHashMap === null) {
+            $valueHashMap = [];
+        }
+
+        $newRecords = [];
+
+        foreach ($records as $_idx => $_record) {
+            $newRecords[$_idx] = $this->_extractRecordData($_record, $valueHashMap);
+        }
+
+        return $newRecords;
+    }
+
+    /**
+     * Extracts record's data from the container and into an array.
+     *
+     * @since [*next-version*]
+     *
+     * @param array|ArrayAccess|stdClass|ContainerInterface $record       The record data container.
+     * @param array                                         $valueHashMap A value-to-hash map reference to which new
+     *                                                                    value-hash pairs are written.
+     *
+     * @throws ContainerExceptionInterface If an error occurred while reading from the record container.
+     *
+     * @return array The extracted record data as an associative array.
+     */
+    protected function _extractRecordData($record, &$valueHashMap = [])
+    {
+        // Initialize variable, in case it was declared implicitly during the method call
+        if ($valueHashMap === null) {
+            $valueHashMap = [];
+        }
+
+        $result = [];
+
+        foreach ($this->_getSqlInsertFieldColumnMap() as $_field => $_column) {
```

Comment:
Would it make more sense to de-couple this method from field-column map awareness? Also, why is it in the same trait? Looks quite generic to me.

Reply:
> As for decoupling from field-column awareness ... I'm not sure what else it could depend on.

Reply:
Oh! That makes sense.